shared-libs/cht-datasource/src/libs/data-context.ts

```diff
@@ -39,3 +39,26 @@ export const adapt = <T>(
   assertRemoteDataContext(context);
   return remote(context);
 };
+
+/** @internal */
+export const getDocumentStream = async function* <T, Args extends unknown[]>(
+  fetchFunction: (...args: Args) => Promise<T[]>,
+  fetchFunctionArgs: {
+    limit: number;
+    skip: number;
+  } & Record<string, unknown>
+): AsyncGenerator<unknown, void> {
+  const { limit } = fetchFunctionArgs;
+  let { skip } = fetchFunctionArgs;
+  const hasMoreResults = () => skip && skip % limit === 0;
+
+  do {
+    const docs = await fetchFunction(...(Object.values(fetchFunctionArgs) as Args));
```

Comment:
I'm not a fan of using `...Object.values(args)` as arguments because it now creates an implicit dependency on the order of the properties. In practice, it means that passing `{ limit: 100, skip: 500 }` will yield different results from `{ skip: 500, limit: 100 }` based on how `fetchFunction` expects parameters to be ordered

Reply:
So I've been mulling over this and I believe we could simplify it by making `fetchFunction` expect a single object argument, that way the ordering doesn't matter and we get a clear type inference

```ts
interface PaginationArgs {
  limit: number;
  skip: number;
}

type DefaultArgs = PaginationArgs & Record<string, unknown>;

/** @internal */
export const getDocumentStream = async function* <T, Args extends DefaultArgs>(
  fetchFunction: (args: Args) => Promise<T[]>,
  fetchFunctionArgs: Args
): AsyncGenerator<unknown, void> {
  const { limit } = fetchFunctionArgs;
  let { skip } = fetchFunctionArgs;
  const hasMoreResults = () => skip && skip % limit === 0;

  do {
    const docs = await fetchFunction(fetchFunctionArgs);

    for (const doc of docs) {
      yield doc;
    }

    skip += docs.length;
  } while (hasMoreResults());
};
```

and `getPeople()` in `src/person.ts` could look like this to accommodate for the different function signature:

```ts
  const getPeople = <T>(
    localFn: (c: LocalDataContext) => (personType: ContactTypeQualifier, limit: number, skip: number) => Promise<T>,
    remoteFn: (c: RemoteDataContext) => (personType: ContactTypeQualifier, limit: number, skip: number) => Promise<T>
  ) => (context: DataContext) => {
      assertDataContext(context);
      const fn = adapt(context, localFn, remoteFn);

      /**
       * Returns an array of people for the provided page specifications.
       * @param params parameters
       * @param params.personType the type of people to return
       * @param params.limit the maximum number of people to return. Default is 100.
       * @param params.skip the number of people to skip. Default is 0.
       * @returns an array of people for the provided page specifications.
       * @throws Error if `personType` qualifier is invalid
       * @throws Error if the provided `limit` value is `<=0`
       * @throws Error if the provided `skip` value is `<0`
       */
      const curriedFn = async ({personType, limit = 100, skip = 0}: {
      personType: ContactTypeQualifier,
      limit: number,
      skip: number
    }): Promise<T> => {
        assertTypeQualifier(personType);
        assertLimitAndSkip(limit, skip);

        return fn(personType, limit, skip);
      };
      return curriedFn;
    };
```

The formatting is very rough but let me know if the idea makes sense